examples/common/optim/adaptive_lion.py

```diff
@@ -0,0 +1,401 @@
+            betas = betas,
+            weight_decay = weight_decay
+        )
+
+        super().__init__(params, defaults)
+        
+        for group in self.param_groups:
+            group['initial_lr'] = group['lr']
+        self.outlier_threshold = outlier_threshold
+        self.timeout = timeout
+        self.lr_penalty = lr_penalty
+        self.min_scale = min_scale
+
+    @staticmethod
+    def lionw(p, grad, exp_avg, lr, initial_lr, wd, beta1, beta2) -> None:
+        # stepweight decay
+        if wd != 0:
+            decay_factor = (lr / initial_lr) if initial_lr else 1.0
+            p.data.mul_(1 - decay_factor * wd)
+
+        # update is interpolation between gradient and momentum
+        update = exp_avg.lerp(grad, 1 - beta1).sign_()
+        p.add_(update, alpha = -lr)
+
+        # momentum is interp b/w gradient and itself
+        exp_avg.lerp_(grad, 1 - beta2)
+
+    @staticmethod
+    def adjust_lr(lr, lr_penalty, num_times, min_scale):
+        return lr * max(min_scale, lr_penalty ** num_times)
+
+    @torch.no_grad()
+    def step(
+        self,
+        closure: Optional[Callable] = None
+    ):
+
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+        
+        for group in self.param_groups:
+            for p in filter(lambda p: p.grad is not None and p.requires_grad, group['params']):
+
+                grad, lr, initial_lr, wd, beta1, beta2, state = p.grad, group['lr'], group['initial_lr'], group['weight_decay'], *group['betas'], self.state[p]
+
+                # init state - exponential moving average of gradient values
+
+                if len(state) == 0:
+                    state['exp_avg'] = torch.zeros_like(p)
+                    state['moment_tracker'] = OutlierDetector(self.outlier_threshold)
+                    state['outlier_timestamp'] = []
+                    state['step'] = 0
+
+                exp_avg = state['exp_avg']
+
+                # determine if the new moment resulting from this grad would be an outlier
+                moment_norm = torch.linalg.vector_norm(
+                    exp_avg.lerp(grad, 1 - beta2)
+                ) ** 2
+
+                if dist.get_world_size() > 1:
+                    dist.all_reduce(moment_norm, reduce_operation='SUM')
+                moment_norm = math.sqrt(moment_norm)
+
+                if state['moment_tracker'].insert_observation(moment_norm):
+                    state['outlier_timestamp'].append(state['step'])
+                
+                removed = []
+                for ts in state['outlier_timestamp']:
+                    if state['step'] - ts > self.timeout:
+                        removed.append(ts)
+                
+                for ts in removed:
+                    state['outlier_timestamp'].remove(ts)
+                
+                lr = self.adjust_lr(lr, self.lr_penalty, len(state['outlier_timestamp']), self.min_scale)
```

Comment:
am i reading this correctly that you basically decrease the learning rate proportional to the number of outliers in the previous `timeout` number of steps? would be good to add an explanation of the algorithm to the docstring